### PR TITLE
[fix] #0000: On-chain predictable iteration order

### PR DESCRIPTION
**Pull request by mversic**
## Description

* replace all `HashMap`/`HashSet` in `iroha_core` with `IndexMap`/`IndexSet`
* add `HashMap`/`HashSet` to `dissallowed_types` 

### Linked issue

Closes [#3](https://app.zenhub.com/workspaces/iroha-v2-60ddb820813b9100181fc060/issues/zh/3)

### Checklist

- [ ] I've read `CONTRIBUTING.md`
- [ ] I've used the standard signed-off commit format (or will squash just before merging)
- [ ] All applicable CI checks pass (or I promised to make them pass later)
- [ ] (optional) I've written unit tests for the code changes
- [ ] I replied to all comments after code review, marking all implemented changes with thumbs up

### CODE REVIEW

File: core/clippy.toml
Comment (mversic) [inline core/clippy.toml:1]: can we agree this is the safest thing to do

Comment (Arjentix) [reply core/clippy.toml:1]: > can we agree this is the safest thing to do?

Yeah, looks like a good feature

> should I do this at workspace level?

I think yes and theh disable in client-facing crates

Comment (Arjentix) [reply core/clippy.toml:1]: Because it might be useful for `smart_contract`, `primitives` and etc

**Review by Arjentix — APPROVED** [review]
In the issue you wrote that you'll use `BTree*` collections. Why did you change your mind?

File: core/clippy.toml
Comment (mversic) [reply core/clippy.toml:1]: > I think yes and theh disable in client-facing crates

it doesn't work. If there are 2 overlapping configs workspace level wins. Check [here](https://github.com/rust-lang/rust-clippy/issues/7353)

**Comment by mversic** [comment]
> In the issue you wrote that you'll use `BTree*` collections. Why did you change your mind?

it seemed as a more natural replacement. As long as the bug is fixed we can always switch to b-trees.

File: core/Cargo.toml [hunk 1]
```diff
@@ -68,7 +68,9 @@ displaydoc = { workspace = true }
 wasmtime = { workspace = true }
 parking_lot = { workspace = true, features = ["deadlock_detection"] }
 derive_more = { workspace = true }
+
```
[the remark this answers is not in the document]
Comment (mversic) [reply core/Cargo.toml:71]: yeah

**Review by Erigara — APPROVED** [review]

**Review by Arjentix — APPROVED** [review]
So index at the end

**Pull request merged**